source/include/ble/cs_UUID.h

```diff
@@ -1,145 +1,127 @@
-/**
- * @file
- * Universally Unique IDentifiers for BLE services and characteristics.
- *
- * @authors Crownstone Team, Christopher Mason.
- * @copyright Crownstone B.V.
- * @date May 4, 2015
+/*
+ * Author: Crownstone Team
+ * Copyright: Crownstone (https://crownstone.rocks)
+ * Date: Apr 7, 2021
  * License: LGPLv3+, Apache License 2.0, and/or MIT (triple-licensed)
  */
+
 #pragma once
 
 #include <ble/cs_Nordic.h>
+#include <protocol/cs_Typedefs.h>
 
-// @TODO Implement in a clean, methodical and understandable way
-
-/** A Universally Unique IDentifier for BLE services and characteristics.
+/**
+ * Class that enables the use of 128 bit service UUIDs.
+ *
+ * There are predefined 16 bit UUIDs, which can be found here:
+ * https://www.bluetooth.com/specifications/assigned-numbers/
+ *
+ * Vendor specific UUIDs are usually 128 bit.
+ *
+ * 128b UUID strings are reverse ordered compared to the byte array, in the form:
+ * "15 14 13 12 - 11 10 - 09 08 - 07 06 - 05 04 03 02 01 00"
+ * but then without spaces.
  *
- * This is a 128-bit sequence for non-standard profiles, services, characteristics, or descriptors.
- * There are several identifiers predefined at
- * https://developer.bluetooth.org/gatt/services/Pages/ServicesHome.aspx.
- * Normally a 16-bit sequence is enough to distinguish the different services, a large part of the 128-bit
- * sequence is repeated.
+ * The Softdevice functions only deal with 16b UUIDS, so they introduced 1 extra byte (the 'type' field) to indicate the base UUID.
+ * This base UUID has to be registered first, but can then be used with different 16b fields.
+ * A base UUID would look like XXXX0000-XXXX-XXXX-XXXX-XXXXXXXXXXXX.
+ * The 0000 part is then replaced by the 16b UUID.
+ *
+ * A typical example is a BLE service with multiple characteristics.
+ * All of them will have the same base UUID, but different 16b UUIDs.
  */
 class UUID {
```

Comment:
Just to get me up to speed:
Who should construct these UUID objects and how shoudl the be used?
What work is done in the constructors?
I'm not sure if I get what the `fromXXX` do. It seems they indirectly register something with the soft device, right?
What happens when a UUID object is destroyed? Do we permanently loose the allocated space in the soft device?